onnxruntime/core/framework/fallback_cpu_capability.cc

```diff
@@ -45,10 +45,16 @@ std::unordered_set<NodeIndex> GetCpuPreferredNodes(const onnxruntime::GraphViewe
     return node_id_to_order_map[n1] > node_id_to_order_map[n2];
   };
 
-  std::priority_queue<NodeIndex, std::vector<NodeIndex>, decltype(greater_order_comp)> candidates(greater_order_comp);
+  // If return false, n2 will be output first; If return true, n1 will be output first
+  auto lesser_order_comp = [&](const NodeIndex n1, const NodeIndex n2) {
+    return node_id_to_order_map[n1] < node_id_to_order_map[n2];
+  };
+
+  std::priority_queue<NodeIndex, std::vector<NodeIndex>, decltype(greater_order_comp)> candidates_fw(greater_order_comp);
+  std::priority_queue<NodeIndex, std::vector<NodeIndex>, decltype(lesser_order_comp)> candidates_bw(lesser_order_comp);
```

Comment:
There is an implicit assumption here: NodeIndex is an approximate ordering for the execution order. 
I think this is a wrong assumption. 
To get the real execution order, we need to do an topological sort. 
Let me construct an offending case for you.

Reply:
![image](https://user-images.githubusercontent.com/9906745/115911085-814bbd80-a422-11eb-8b68-40feab3b0a90.png)

Reply:
In the forward pass, by the time we visit Concat node, the right input is not yet placed in cpu_args yet... 
As a result, concat node would not be placed on CPU.

Reply:
Let cover such cases of  "multi-starting point, multi-ending point" in the UT.

Reply:
This is something I learnt during the Cast Propagation work. 
The algorithm we are we are taking here is "CPU colored-subgraph" propagation. 
And to propagate the "frontier" of a colored-subgraph, we should always use topological order.

Reply:
Topological ordering is already in use in the comparator function : https://github.com/microsoft/onnxruntime/blob/331cd1c824b885447b13ea8656d2674e51b002b3/onnxruntime/core/framework/fallback_cpu_capability.cc#L36
I have added the test you specified as well.